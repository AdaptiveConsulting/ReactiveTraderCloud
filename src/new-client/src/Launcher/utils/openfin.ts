// TODO - Revisit if/when launcher is cross platform
import { Application } from "openfin/_v2/main"
import { Bounds } from "openfin/_v2/shapes/shapes"
import { useEffect } from "react"

export async function getExistingOpenFinApplication(
  uuid: string,
): Promise<Application | undefined> {
  const allApps = await fin.System.getAllApplications()
  const targetApp = allApps.some((app) => app.uuid === uuid)
  if (targetApp) {
    return fin.Application.wrap({ uuid })
  }
}

async function restoreExistingApp(existingApp: Application): Promise<void> {
  const isRunning = await existingApp.isRunning()
  if (!isRunning) {
    await existingApp.run()
    return
  }

  const window = await existingApp.getWindow()
  await window.restore()
  await window.bringToFront()
}

export async function bringToFrontOpenFinApplication(
  uuid: string,
): Promise<Application | undefined> {
  const existingApp = await getExistingOpenFinApplication(uuid)
  if (existingApp) {
    await restoreExistingApp(existingApp)
    return existingApp
  }
}

export async function createAndRunOpenFinApplication(
  name: string,
  url: string,
  uuid?: string,
  windowOptions?: fin.WindowOption,
): Promise<Application> {
  const appOptions: fin.ApplicationOption = {
    name,
    url,
    uuid: uuid || name,
    nonPersistent: true,
    mainWindowOptions: windowOptions,
    fdc3Api: true,
  }

  return fin.Application.start(appOptions)
}

export function createOpenFinWindow(
  name: string,
  url: string,
  windowOptions?: fin.WindowOption,
): Promise<fin.OpenFinWindow> {
  return new Promise((resolve, reject) => {
    const window: fin.OpenFinWindow = new fin.desktop.Window(
      {
        url,
        name,
        ...windowOptions,
      },
      () => resolve(window),
      reject,
    )
  })
}

export const minimiseCurrentWindow = async () => {
  const window = await fin.Window.getCurrent()
  window.minimize()
}

export const getCurrentWindowBounds = async () => {
  const window = await fin.Window.getCurrent()
  return window.getBounds()
}

export async function animateCurrentWindowSize(
  bounds: Bounds,
  duration: number = 200,
) {
  const window = await fin.Window.getCurrent()

  return window.animate(
    {
      size: {
        duration,
        height: bounds.height,
        width: bounds.width,
      },
    },
    {
      tween: "ease-in-out",
      interrupt: true,
    },
  )
}

export function useAppBoundReset(bounds: Bounds | undefined) {
  useEffect(() => {
    if (!bounds) {
      return
    }

    const resetAppBound = () => {
      animateCurrentWindowSize({
        ...bounds,
      })
    }

    window.addEventListener("beforeunload", resetAppBound)

    return () => {
      window.removeEventListener("beforeunload", resetAppBound)
    }
  }, [bounds])
}
