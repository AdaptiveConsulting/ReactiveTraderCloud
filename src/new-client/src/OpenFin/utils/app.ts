export function getEnvironment(): string {
  const serviceUrl = window.location.host
  let envMatch

  if (serviceUrl.includes("localhost")) {
    return "local"
  }

  if (serviceUrl === "www.reactivetrader.com") {
    return "demo"
  }

  envMatch = /^(?<env>(?:dev|uat))\.reactivetrader\.com/.exec(serviceUrl)

  if (envMatch?.groups) {
    return envMatch.groups.env
  }

  envMatch = /web-(?<env>\w+)\.adaptivecluster\.com/.exec(serviceUrl)

  if (envMatch?.groups) {
    return envMatch.groups.env
  }

  envMatch = /^(?<env>\w+)\.lb\.adaptivecluster\.com/.exec(serviceUrl)

  if (envMatch?.groups) {
    return envMatch.groups.env
  }

  return "unknown"
}

const prodEnvs = ["demo"]

export function getAppName(name: string = "Reactive TraderÂ®"): string {
  const env = getEnvironment() || "unknown"
  const envFormatted = prodEnvs.includes(env) ? "" : ` (${env.toUpperCase()})`

  return `${name}${envFormatted}`
}

/**
 * Gets the title of the first `content` element of the current Platform Layout
 */
export async function getPlatformLayoutTitle(): Promise<string | undefined> {
  const layout = fin.Platform.Layout.getCurrentSync()
  const config = await layout.getConfig()
  return config.content?.[0].title
}
