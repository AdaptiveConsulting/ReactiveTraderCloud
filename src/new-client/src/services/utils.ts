export type CamelCase<T> = { [P in keyof T & string as Uncapitalize<P>]: T[P] }

export interface CollectionUpdates {
  IsStateOfTheWorld: boolean
  IsState: boolean
}

export enum UpdateType {
  Added = "Added",
  Removed = "Removed",
}

export interface CollectionUpdate {
  UpdateType: UpdateType
}
