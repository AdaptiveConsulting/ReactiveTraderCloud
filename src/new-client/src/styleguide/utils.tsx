const NOTIONALS_STORAGE_KEY = "notionals"

export const setNotionalOnStorage = (
  currencyPairSymbol: string,
  value: number,
) => {
  if (localStorage) {
    const notionals = localStorage.getItem(NOTIONALS_STORAGE_KEY)
    const newNotionalsValue =
      notionals !== null
        ? { ...JSON.parse(notionals), [currencyPairSymbol]: `${value}` }
        : { [currencyPairSymbol]: `${value}` }

    localStorage.setItem(
      NOTIONALS_STORAGE_KEY,
      JSON.stringify(newNotionalsValue),
    )
  }
}
