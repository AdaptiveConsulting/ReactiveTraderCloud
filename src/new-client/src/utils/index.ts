export * from "./formatNumber"
export * from "./equals"
export * from "./mapObject"
export * from "./usePopUpMenu"
export * from "./capitalize"
export * from "./windowCoordinates"
export * from "./isMobile"
export * from "./isPwa"
export * from "./useLocalStorage"
