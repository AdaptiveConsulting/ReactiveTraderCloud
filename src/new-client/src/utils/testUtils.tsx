import { Component, ErrorInfo } from "react"
import { ThemeProvider } from "@/theme"

class LocalStorageMock {
  store: Record<string, string>
  constructor() {
    this.store = {}
  }

  clear() {
    this.store = {}
  }

  getItem(key: string) {
    return this.store[key] || null
  }

  setItem(key: string, value: string) {
    this.store[key] = value
  }

  removeItem(key: string) {
    delete this.store[key]
  }

  get length(): number {
    return Object.keys(this.store).length
  }
  key() {
    return null
  }
}

const mockedLocalStorage = new LocalStorageMock()

export const TestThemeProvider: React.FC = ({ children }) => (
  <ThemeProvider storage={mockedLocalStorage}>{children}</ThemeProvider>
)

export class TestErrorBoundary extends Component<
  {
    onError: (error: Error, errorInfo: ErrorInfo) => void
  },
  {
    hasError: boolean
  }
> {
  state = {
    hasError: false,
  }

  componentDidCatch(error: Error, errorInfo: ErrorInfo) {
    this.props.onError(error, errorInfo)
  }

  static getDerivedStateFromError() {
    return { hasError: true }
  }

  render() {
    if (this.state.hasError) {
      return "error"
    }

    return this.props.children
  }
}
