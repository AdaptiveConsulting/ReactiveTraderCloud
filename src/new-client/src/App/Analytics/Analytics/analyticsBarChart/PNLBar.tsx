import React, { useState } from "react"
import { formatNumber } from "utils/formatNumber"
import {
  BarChart,
  BarPriceContainer,
  Label,
  Offset,
  PriceContainer,
  PriceLabel,
  DiamondShape,
  Bar,
  OriginTickWrapper,
  OriginTick,
} from "./styled"
interface PNLBarProps {
  basePnl: number
  maxVal: number
  symbol: string
}

const TRANSLATION_WIDTH: number = 50

const getLogRatio: (max: number, numb: number) => number = (max, numb) => {
  const logMax = Math.log10(Math.abs(max)) + 1
  const logNumb = Math.log10(Math.abs(numb))
  return logNumb / logMax
}

const PNLBar: React.FC<PNLBarProps> = ({ symbol, basePnl, maxVal }) => {
  const [hovering, setHovering] = useState(false)
  const color = basePnl >= 0 ? "positive" : "negative"
  const distance =
    getLogRatio(maxVal, basePnl) * TRANSLATION_WIDTH * (basePnl >= 0 ? 1 : -1)
  const price = formatNumber(Math.abs(basePnl))
  const hoverPrice = formatNumber(basePnl)
  return (
    <BarChart>
      <Label>{symbol}</Label>
      <Offset />
      <BarPriceContainer>
        <PriceContainer distance={distance}>
          <PriceLabel
            onMouseEnter={() => setHovering(true)}
            onMouseLeave={() => setHovering(false)}
            distance={distance}
            color={color}
          >
            {hovering ? hoverPrice : price}
          </PriceLabel>
          <DiamondShape color={color} />
        </PriceContainer>
        <Bar />
        <OriginTickWrapper>
          <OriginTick />
        </OriginTickWrapper>
      </BarPriceContainer>
    </BarChart>
  )
}

export default PNLBar
