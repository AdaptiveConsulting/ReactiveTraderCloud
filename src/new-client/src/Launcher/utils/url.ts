import { BASE_URL, ENVIRONMENT } from "@/constants"
import { constructUrl } from "@/utils/url"

export const getReactiveTraderUrl = (path: string) =>
  constructUrl(
    path,
    ENVIRONMENT !== "local"
      ? BASE_URL.replace("launcher", "openfin")
      : "http://localhost:1917",
  )
