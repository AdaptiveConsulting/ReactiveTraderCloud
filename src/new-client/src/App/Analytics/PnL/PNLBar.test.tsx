import { render, screen, act, fireEvent } from "@testing-library/react"
import { TestThemeProvider } from "utils/testUtils"
import renderer from "react-test-renderer"
import PNLBar from "./PNLBar"

const renderComponent = (symbol: string, basePnl: number, maxVal: number) =>
  render(
    <TestThemeProvider>
      <PNLBar symbol={symbol} basePnl={basePnl} maxVal={maxVal} />
    </TestThemeProvider>,
  )

describe("Tile", () => {
  it("should display the symbol name correctly", () => {
    renderComponent("EURAUD", -26043.691207338878, 1239939.1231111237)
    expect(screen.getByTestId("symbolLabel").textContent).toBe(`EURAUD`)
  })

  it("should display the price and hover price correctly", () => {
    renderComponent("EURAUD", -26043.691207338878, 1239939.1231111237)
    expect(screen.getByTestId("priceLabel").textContent).toBe(`-26k`)

    act(() => {
      fireEvent.mouseOver(screen.getByTestId("priceLabel"))
    })

    expect(screen.getByTestId("priceLabel").textContent).toBe(`-26,043.69`)
  })

  it("renders the bar correctly", () => {
    const tree = renderer
      .create(
        <TestThemeProvider>
          <PNLBar
            symbol={"EURAUD"}
            basePnl={-26043.691207338878}
            maxVal={1239939.1231111237}
          />
        </TestThemeProvider>,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
