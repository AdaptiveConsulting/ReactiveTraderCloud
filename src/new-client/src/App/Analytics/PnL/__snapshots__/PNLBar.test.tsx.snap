// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tile renders the bar correctly 1`] = `
<div
  className="styled__BarChart-sc-3sau7y-1 fLKfIL"
>
  <div
    className="styled__Label-sc-3sau7y-7 IDAmG"
    data-testid="symbolLabel"
  >
    EURAUD
  </div>
  <div
    className="styled__Offset-sc-3sau7y-3 bbLujG"
  />
  <div
    className="styled__BarPriceContainer-sc-3sau7y-8 bsBpLt"
  >
    <div
      className="styled__FlexDiv-sc-3sau7y-0 styled__PriceContainer-sc-3sau7y-2 glNClZ llriux"
      style={
        Object {
          "transform": "translate(-31.125428646769944%)",
        }
      }
    >
      <div
        className="styled__PriceLabel-sc-3sau7y-5 hqeAlp"
        color="negative"
        data-testid="priceLabel"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        -26k
      </div>
      <div
        className="styled__DiamondShape-sc-3sau7y-6 Usval"
        color="negative"
      />
    </div>
    <div
      className="styled__Bar-sc-3sau7y-9 eBLUMA"
    />
    <div
      className="styled__FlexDiv-sc-3sau7y-0 styled__OriginTickWrapper-sc-3sau7y-4 glNClZ eGNFGg"
    >
      <div
        className="styled__OriginTick-sc-3sau7y-10 hFQYTC"
      />
    </div>
  </div>
</div>
`;
